engine/client/cl_spray.c

```diff
@@ -0,0 +1,442 @@
+
+	palette_offset = miptex.offsets[3] + mip3_size;
+	palette_size = SPRAY_PALETTE_SIZE;
+
+	FS_Write( f, &header, sizeof( header ));
+	FS_Write( f, &miptex, sizeof( mip_t ));
+	FS_Write( f, indexed, mip0_size );
+	FS_Write( f, mip1_data, mip1_size );
+	FS_Write( f, mip2_data, mip2_size );
+	FS_Write( f, mip3_data, mip3_size );
+	FS_Write( f, &palette_size, sizeof( short ));
+	FS_Write( f, palette, SPRAY_PALETTE_BYTES );
+
+	// padding up to a multiple of 4
+	pad = (( FS_Tell( f ) + 3 ) & ~3 ) - FS_Tell( f );
+	for( i = 0; i < pad; ++i )
+	{
+		FS_Write( f, &zero, 1 );
+	}
+
+	infotableofs = FS_Tell( f );
+	memset( &lump, 0, sizeof( lump ));
+	lump.filepos = sizeof( wad3header_t );
+	lump.disksize = (int)( palette_offset + sizeof( short ) + SPRAY_PALETTE_BYTES );
+	lump.size = lump.disksize;
+	lump.type = (char)lump_type;
+	lump.compression = 0;
+	strncpy( lump.name, SPRAY_NAME, sizeof( lump.name ) - 1 );
+	FS_Write( f, &lump, sizeof( lump ));
+
+	FS_Seek( f, offsetof( wad3header_t, infotableofs ), SEEK_SET );
+	infotableofs32 = (int)infotableofs;
+	FS_Write( f, &infotableofs32, sizeof( int ));
+
+	FS_Close( f );
+	result = true;
+
+cleanup:
+	if( mip1_data )
+		Mem_Free( mip1_data );
+	if( mip2_data )
+		Mem_Free( mip2_data );
+	if( mip3_data )
+		Mem_Free( mip3_data );
+	return result;
+}
+
+// convert BMP to spray WAD3
+static qboolean CL_ConvertBMPToSprayWAD3( const char *logoname, const char *logocolor )
+{
+	char      spray_bmp_path[MAX_QPATH];
+	rgbdata_t *image;
+	int       width, height;
+	rgbdata_t *scaled_image = NULL;
+	byte      palette[SPRAY_PALETTE_BYTES];
+	const logo_color_t *color;
+	float     t;
+	int       i;
+	qboolean  result;
+
+	Q_snprintf( spray_bmp_path, sizeof( spray_bmp_path ), "logos/%s.bmp", ( logoname && *logoname ) ? logoname : "lambda" );
+
+	Image_SetForceFlags( IL_KEEP_8BIT );
+	image = FS_LoadImage( spray_bmp_path, NULL, 0 );
+	Image_SetForceFlags( 0 );
+	if( !image || !image->palette || !( image->type == PF_INDEXED_24 || image->type == PF_INDEXED_32 ))
+	{
+		Con_DPrintf( S_ERROR "Failed to load indexed BMP: %s\n", spray_bmp_path );
+		if( image )
+			FS_FreeImage( image );
+		return false;
+	}
+
+	width = image->width;
+	height = image->height;
+	CL_AdjustSprayDimensions( &width, &height );
+	if( width < 16 || height < 16 || width % 16 != 0 || height % 16 != 0 )
+	{
+		Con_DPrintf( S_ERROR "Invalid BMP dimensions after adjustment: %dx%d\n", width, height );
+		FS_FreeImage( image );
+		return false;
+	}
+
+	if( width != image->width || height != image->height )
+	{
+		rgbdata_t *tmp = CL_LoadAndPrepareImage( spray_bmp_path, &width, &height, &scaled_image );
+		if( !tmp )
+		{
+			Con_DPrintf( S_ERROR "Failed to resample BMP for spray\n" );
+			FS_FreeImage( image );
+			return false;
+		}
+		if( tmp != image )
+			FS_FreeImage( tmp );
+		if( scaled_image )
+		{
+			FS_FreeImage( image );
+			image = scaled_image;
+		}
+	}
+
+	// generate gradient palette
+	color = LogoColor_FindByName( logocolor );
```

Comment:
You don't have to do this. If you take `remapped.bmp`, palette can be taken from that file. 

`imagelib` can preserve palette and won't transform indexed image to RGBA with `KEEP_8BIT` flag.

As for `remapped.png`, it's always fullcolor and can be quantized into indexed image (we also have quantizer in `imagelib`).